### ISSUE TITLE
[MINSTALL-96] Skip option for install-file

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-install-plugin#154]
**[Tiemo Vorschütz](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tiemo)** opened **[MINSTALL-96](https://issues.apache.org/jira/browse/MINSTALL-96?redirect=false)** and commented

For my process it would be helpful to also introduce a skip option for install:install-file to be able to disable install-file creation by setting a pom property.

This could be necessary if the build chain is designed by a parent pom and should by simply modified inside the child pom.

In my case all plugin definitions are outsourced to the parent pom file and should only be set by properties by the developers to take care that modifications to the build chain are inherited.

I attached a patch to enable skipping also for install-file.

---

**Affects:** 2.4

**Attachments:**
- [InstallFileMojo.java.patch](https://issues.apache.org/jira/secure/attachment/12716594/InstallFileMojo.java.patch) (_1.07 kB_)

1 votes, 4 watchers

**Comment by jira-importer** [comment]
**[Ivan Bondarenko](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=buzzeri)** commented

The same need here, but for a bit different task.
We need to run install-file on internal properties condition. Currently we can only put intall-file into profile (only external parameter condition will work) or perform operation in any case. We have chosen later as lesser of two evils, but now some artifacts are installed twice (not a good practice).

**Comment by jira-importer** [comment]
**[Ivan Bondarenko](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=buzzeri)** commented

1) Global aim we want to achieve is following: all artifacts must be installed to local repository. As for deploying to remote repository, in some cases we want to deploy sources, in some we don't. If we don't, we just not attaching 'sources' artifact and install it with 'install-file'. If we do want, we install with 'install' and switch off 'install-file' (what we cannot do right now). In fact, some 'un-attach-artifact' mojo may be better practice here, but there is not such plugin. So 'skip' parameter in 'install-file' will be quite good solution.
2) In fact if my will, I would make 'skip' and maybe some other parameters mandatory for all mojos.
3) 'skip' will be disabled by default, so users may not even know about new change.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

1) The `unattach-artifact` is not the best practice. If you don't want the attachment, then don't attach it. Otherwise the project will be in a wrong state between 2 goals. Good maven-plugins don't have a delete or detach-option, but rather an exclude. I hope you're aware that the maven-source-plugin has an [attach](http://maven.apache.org/plugins/maven-source-plugin/jar-mojo.html#attach) parameter, which you can set to `false`
   2+3) I know there's an issue for it, but I don't think it will work. First of all, `Mojo` is the interface which every goal has to implement. If we would add a skip-method, a lot of plugins will probably break. Even though there is an `AbstractMojo`, plugins are not required to use it. So Maven would have to use reflection to solve this issue. The next problem is that the skip-parameter must have a unique property/expression, otherwise you would be skipping nothing or everything. So still every `Mojo` has to implement some skip-logic.

> all artifacts must be installed to local repository
> Do you mean that every  developer must run `mvn install` on his system?

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

You've just described a perfect example where you thought that skipping would be the best solution, but which would be the start of a chain of other adjustments.
So actually, the `install` phase was always correct: the plugin installed the artifact and the attachments. However, during `deploy` in some cases you don't want to upload the attachments. By trying to solve this issue, the closest you could get was an `install-file` hack. But the hack is not good enough so the hack needs to be hacked.
Instead, I'm willing to do the following: improve the `deploy` goal the a `excludeClassifiedArtifacts` (or something like that) where you can specify the classifiers (and/or extension?) of the attachments which should be excluded.
In conclusion: I still think that the `install-file` goal should not have the `skip` option.

**Comment by jira-importer** [comment]
**[Ivan Bondarenko](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=buzzeri)** commented

Yes, 'skip' is more like a workaround. Excluding certain classifiers (maybe in both install and deploy) is better. Though I'm not sure how it should look like, because different conditional/dynamic stuff is a weak side of Maven. For example "\<excludeTag>sources tests\</excludeTag>" or "\<excludesTag>\<excludeTag>sources\</excludeTag>\<excludeTag>tests\</excludeTag>\</excludesTag>" will not fit because exact classifiers may be not known beforehand. For war-plugin we are solving it with "\<webResources combine.children='append'>", however I don't know is 'combine.children' some common Maven stuff or just related to 'webResources'.
